File: lw7/slides/shapes/Ellipse.cpp
#include "Ellipse.h"

CEllipse::CEllipse(const PointD& center, double horizontalRadius, double verticalRadius)
	: m_center(center)
	, m_horizontalRadius(horizontalRadius)
	, m_verticalRadius(verticalRadius)
{
}

RectD CEllipse::GetFrame()
{
	return RectD{
		m_center.x - m_horizontalRadius,
		m_center.y - m_verticalRadius,
		m_horizontalRadius * 2,
		m_verticalRadius * 2
	};
}

void CEllipse::SetFrame(const RectD& rect)
{
	m_center = PointD{ rect.left + rect.width / 2, rect.top + rect.height / 2 };
	m_horizontalRadius = rect.width / 2;
	m_verticalRadius = rect.height / 2;
}

void CEllipse::DrawBehaviour(ICanvas& canvas) const
{
	PointD leftTop = { m_center.x - m_horizontalRadius, m_center.y - m_verticalRadius };
	double width = 2 * m_horizontalRadius;
	double height = 2 * m_verticalRadius;

	canvas.DrawEllipse(leftTop, width, height);
}

File: lw9/src/model/shapes/Ellipse.js
import {Shape} from "./Shape"
import {Rect} from "../types/Rect.js"
import {Point} from "../types/Point.js"

class Ellipse extends Shape {
    /**
     * @param {Rect} rect
     */
    constructor(rect) {
        super()

        /**
         * @type {number}
         * @private
         */
        this._horizontalRadius = rect.width / 2

        /**
         * @type {number}
         * @private
         */
        this._verticalRadius = rect.height / 2

        /**
         * @type {Point}
         * @private
         */
        this._center = { x: rect.left + this._horizontalRadius, y: rect.top + this._verticalRadius }
    }

    /**
     * @override
     * @private
     */
    _getTypeImpl() {
        return 'ellipse'
    }

    /**
     * @override
     * @private
     */
    _getRectImpl() {
        const left = this._center.x - this._horizontalRadius
        const top = this._center.y - this._verticalRadius

        const width = this._horizontalRadius * 2
        const height = this._verticalRadius * 2

        return {
            left,
            top,
            width,
            height,
        }
    }

    /**
     * @override
     * @private
     */
    _setRectImpl(rect) {
        this._center = { x: rect.left + rect.width / 2, y: rect.top + rect.height / 2 }
        this._horizontalRadius = rect.width / 2
        this._verticalRadius = rect.height / 2
    }
}

export {
    Ellipse,
}